File: ios/UniPark/UniPark/domain/usecases/ScanQRUseCase.swift
import Foundation

public struct ScanQRUseCase {
	private let scanRepository: ScanRepository

	public init(scanRepository: ScanRepository) {
		self.scanRepository = scanRepository
	}

	// MARK: - Execution

	public func execute(
		passPayload: String,
		passSignature: String,
		direction: ScanDirection,
		lotId: UUID
	) async throws -> Scan {
		let idempotencyKey = UUID().uuidString

		do {
			return try await scanRepository.submitScan(
				passPayload: passPayload,
				passSignature: passSignature,
				direction: direction,
				lotId: lotId,
				idempotencyKey: idempotencyKey
			)
		} catch {
			try await scanRepository.syncPendingScans()
			throw error
		}
	}
}

File: android/app/src/main/java/com/unipark/android/domain/usecases/ScanQRUseCase.kt
package com.unipark.android.domain.usecases

import com.unipark.android.domain.entities.Scan
import com.unipark.android.domain.entities.ScanDirection
import com.unipark.android.domain.repositories.ScanRepository
import java.util.UUID
import javax.inject.Inject

class ScanQRUseCase @Inject constructor(
    private val scanRepository: ScanRepository,
) {
    suspend fun execute(
        passPayload: String,
        passSignature: String,
        direction: ScanDirection,
        lotId: UUID,
    ): Scan = scanRepository.submitScan(passPayload, passSignature, direction, lotId)
}
